ipaserver/plugins/baseuser.py

```diff
@@ -728,3 +739,138 @@ def post_callback(self, ldap, dn, entry_attrs, *keys, **options):
+        X509:<I>issuer<S>subject
+        For instance:
+        X509:<I>O=DOMAIN,CN=Certificate Authority<S>O=DOMAIN,CN=user
+        A list of values can be returned if --usercertificate is used multiple
+        times, or in conjunction with --subject --issuer.
+        """
+        data = []
+        data.extend(certmapdata)
+
+        if 'issuer' in options or 'subject' in options:
+            issuer = options.get('issuer')
+            subject = options.get('subject')
+            data.append(ModCertMapData._build_mapdata(subject, issuer))
+
+        for dercert in options.get('usercertificate', ()):
+            cert = x509.load_certificate(dercert, x509.DER)
+            issuer = DN(cert.issuer)
+            subject = DN(cert.subject)
+            data.append(ModCertMapData._build_mapdata(subject, issuer))
+
+        return data
+
+    def get_args(self):
+        # ipacertmapdata is not mandatory as it can be built
+        # from the values subject+issuer or from reading usercertificate
+        for arg in super(ModCertMapData, self).get_args():
+            if arg.name == 'ipacertmapdata':
+                yield arg.clone(required=False, alwaysask=False)
+            else:
+                yield arg.clone()
+
+    def pre_callback(self, ldap, dn, entry_attrs, attrs_list, *keys,
+                     **options):
+        # The 3 valid calls are
+        # ipa user-add-certmapdata LOGIN --subject xx --issuer yy
+        # ipa user-add-certmapdata LOGIN [DATA] --certificate xx
+        # ipa user-add-certmapdata LOGIN DATA
+        # Check that at least one of the 3 formats is used
+
+        # If only LOGIN is supplied, then we need either subject or issuer or
+        # certificate
+        if len(keys) < 2:
+            valid_options = ('issuer', 'subject', 'usercertificate')
+            if all(key not in options for key in valid_options):
+                raise errors.RequirementError(name='CERTMAPDATA')
+
+        # If subject or issuer is provided, other options are not allowed
+        if any(key in options for key in ['issuer', 'subject']):
+            if 'usercertificate' in options:
+                raise errors.MutuallyExclusiveError(
+                    reason=_('cannot specify both subject/issuer '
+                             'and certificate'))
+            if len(keys) >= 2:
+                raise errors.MutuallyExclusiveError(
+                    reason=_('cannot specify both subject/issuer '
+                             'and CERTMAPDATA'))
+
+        # if the command is called with --subject --issuer or --certificate
+        # we need to add ipacertmapdata to the attrs_list in order to
+        # display the resulting value in the command output
+        if 'ipacertmapdata' not in attrs_list:
+            attrs_list.append('ipacertmapdata')
+
+        data = entry_attrs.get(self.attribute, list())
+        entry_attrs[self.attribute] = self._convert_options_to_certmap(
+            data, options)
```

Comment:
IMO it would be better to hide all this inside `_convert_options_to_certmap()` and have it called like this here:
```python
        self._convert_options_to_certmap(entry_attrs, options)
```